### PR TITLE
feat(notifications): Nudge Notifications

### PR DESCRIPTION
**Pull request by mgaeta**
When a user links their identity but doesn't have their notification settings "on" for Slack, send them a reminder message.
Here are some examples:
Screen Shot 2021-12-09 at 10 56 29 AM

### CODE REVIEW

File: src/sentry/integrations/slack/notifications.py [hunk 1]
```diff
@@ -52,7 +55,7 @@ def get_attachments(
 
 
 def get_context(
-    notification: OrganizationNotification,
+    notification: BaseNotification,
```
Comment (ceorourke) [inline src/sentry/integrations/slack/notifications.py:58]: A Slack integration is always related to an organization, so why is this changing?

File: src/sentry/integrations/slack/views/link_identity.py [hunk 2]
```diff
@@ -74,6 +75,10 @@ def handle(self, request: Request, signed_params: str) -> Response:
             Identity.objects.reattach(idp, params["slack_id"], request.user, defaults)
 
         send_slack_response(integration, SUCCESS_LINKED_MESSAGE, params, command="link")
+        if not NotificationSetting.objects.has_any_provider_settings(
```
Comment (ceorourke) [inline src/sentry/integrations/slack/views/link_identity.py:78]: I guess this is happening right after they link their identity, but in what case would a user have Slack settings at this point? If they had unlinked and are re-linking?

File: src/sentry/notifications/notifications/integration_nudge.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+from sentry.notifications.utils import send_base_notification
+from sentry.notifications.utils.actions import MessageAction
+from sentry.types.integrations import ExternalProviders
+
+if TYPE_CHECKING:
+    from sentry.models import Organization, Team, User
+
+logger = logging.getLogger(__name__)
+
+MESSAGE_LIBRARY = [
```
Comment (ceorourke) [inline src/sentry/notifications/notifications/integration_nudge.py:18]: Is there any way for us to know which message(s) was more effective than another?

**Review by ceorourke — COMMENTED** [review]
Can you add a screenshot of what the notification looks like and add tests?

File: src/sentry/integrations/slack/notifications.py [hunk 4]
```diff
@@ -154,52 +207,22 @@ def send_notification_as_slack(
...
-        extra_context = (extra_context_by_actor_id or {}).get(recipient.actor_id, {})
-        context = get_context(notification, recipient, shared_context, extra_context)
-        attachments = get_attachments(notification, recipient, context)
-
+    for recipient, tokens_by_channel in get_channel_and_token_by_recipient(
+        organization=notification.organization,
+        recipients=recipients,
+    ).items():
+        is_multiple = check_if_is_multiple(notification, recipient, tokens_by_channel)
+        attachments = get_attachments(
```
Comment (scefali) [inline src/sentry/integrations/slack/notifications.py:215]: @mgaeta wondering why this change is part of the PR, doesn't seem related to the description unless I'm missing something

File: src/sentry/integrations/slack/notifications.py [hunk 5]
```diff
@@ -142,6 +143,58 @@ def get_channel_and_token_by_recipient(
...
+    notification.record_notification_sent(recipient, ExternalProviders.SLACK)
+
+
+def check_if_is_multiple(
+    notification: BaseNotification,
+    recipient: Team | User,
+    tokens_by_channel: Mapping[str, str],
+) -> bool:
+    is_multiple = len(tokens_by_channel.keys()) > 1
+    if is_multiple:
```
Comment (scefali) [inline src/sentry/integrations/slack/notifications.py:187]: @mgaeta wondering why we have all this logic for `is_multiple`. Is this doing anything functional? Or is it just sued for logging?

Comment (mgaeta) [reply src/sentry/integrations/slack/notifications.py:190]: Yes `is_multiple` is just for logging.

File: src/sentry/integrations/slack/views/link_identity.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/integrations/slack/views/link_identity.py:78]: Yeah if the user has settings and then unlinks and relinks they won't get the notifications. Also if a user links a second identity somehow they could have notification settings.

File: src/sentry/notifications/notifications/integration_nudge.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/notifications/notifications/integration_nudge.py:17]: My plan was to just run the experiments far enough away from each other that we can just see when the numbers go up the fastest.

File: src/sentry/integrations/slack/notifications.py (hunk 4 above)
Comment (ceorourke) [reply src/sentry/integrations/slack/notifications.py:211]: I agree, is it necessary to rewrite so much of this file?

File: src/sentry/integrations/slack/endpoints/action.py [hunk 6]
```diff
@@ -42,6 +50,7 @@
 NO_ACCESS_MESSAGE = "You do not have access to the organization for the invitation."
 NO_PERMISSION_MESSAGE = "You do not have permission to approve member invitations."
 NO_IDENTITY_MESSAGE = "Identity not linked for user."
+ENABLE_SLACK_SUCCESS_MESSAGE = "Slack has been enabled."
```
Comment (ceorourke) [inline src/sentry/integrations/slack/endpoints/action.py:53]: Maybe this could be like "Slack notifications have been enabled" because as it is, it reads a bit strangely since the user obviously came from Slack.

**Review by ceorourke — COMMENTED** [review]
I'd still like to see tests for nudge notifications

File: src/sentry/notifications/notifications/integration_nudge.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        super().__init__(organization)
+        self.recipient = recipient
+        self.provider = provider
+        self.seed = (
+            (seed % len(MESSAGE_LIBRARY))
+            if seed is not None
+            else random.randint(0, len(MESSAGE_LIBRARY) - 1)
+        )
+
+    def get_reference(self) -> Model | None:
```
Comment (leeandher) [inline src/sentry/notifications/notifications/integration_nudge.py:65]: Are most of these methods just required from the base class? Is that why they just return empty values?

**Review by leeandher — APPROVED** [review]

**Comment by leeandher** [comment]
There are some broken tests right now and nudge notifications still need coverage, but the rest looks good to me

File: src/sentry/notifications/notifications/integration_nudge.py (hunk 7 above)
Comment (mgaeta) [reply src/sentry/notifications/notifications/integration_nudge.py:65]: Yeah they're used in constructing the email version of the message.

File: src/sentry/notifications/notifications/integration_nudge.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+class IntegrationNudgeNotification(BaseNotification):
+    category = "integration_nudge"
+    filename = "integration-nudge"
+    type = "integration.nudge"
+
+    def __init__(
+        self,
+        organization: Organization,
+        recipient: User,
+        provider: ExternalProviders,
```
Comment (scefali) [inline src/sentry/notifications/notifications/integration_nudge.py:53]: @mgaeta Not understanding why this is a parameter to the class. It looks like this won't work for email. Can you just implement this logic in `get_notification_providers` instead?

File: src/sentry/notifications/notifications/integration_nudge.py (hunk 3 above)
Comment (scefali) [reply src/sentry/notifications/notifications/integration_nudge.py:17]: @mgaeta I would put this info in the analytics event

File: src/sentry/notifications/notifications/integration_nudge.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    def get_notification_title(self) -> str:
+        return ""
+
+    def get_title_link(self) -> str | None:
+        return None
+
+    def build_attachment_title(self) -> str:
+        return ""
+
+    def get_filename(self) -> str:
```
Comment (scefali) [inline src/sentry/notifications/notifications/integration_nudge.py:98]: @mgaeta Wondering why all these functions return an empty string. IMO it's better to let it error out because any behavior coming from this would be invalid

File: src/sentry/notifications/manager.py [hunk 10]
```diff
@@ -389,3 +390,34 @@ def disable_settings_for_users(
                     target_id=user.actor_id,
                     defaults={"value": NotificationSettingOptionValues.NEVER.value},
                 )
+
+    def has_any_provider_settings(
+        self, recipient: Team | User, provider: ExternalProviders
+    ) -> bool:
+        # Explicitly typing to satisfy mypy.
+        has_settings: bool = (
+            self._filter(provider=provider, target_ids={recipient.actor_id})
```
Comment (scefali) [inline src/sentry/notifications/manager.py:399]: @mgaeta How does this logic know to check a user vs a team? Couldn't the IDs collide?

File: src/sentry/notifications/notifications/integration_nudge.py (hunk 7 above)
Comment (scefali) [reply src/sentry/notifications/notifications/integration_nudge.py:65]: @mgaeta right...but why do we need to define them if we aren't sending an email?

File: src/sentry/notifications/notifications/integration_nudge.py (hunk 8 above)
Comment (mgaeta) [reply src/sentry/notifications/notifications/integration_nudge.py:53]: I built the notification to be provider-agnostic because we'll need another nudge every time we add a _new_ provider.

File: src/sentry/notifications/manager.py (hunk 10 above)
Comment (mgaeta) [reply src/sentry/notifications/manager.py:399]: The actor ID space is shared by users and teams.
